{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/lij/PycharmProjects/eclip_analysis_pipeline/20201027_pileup_data/pileup_output/18s_pileup_table.csv\n",
      "Detected 18s as contig used in filename\n",
      "/Users/lij/PycharmProjects/eclip_analysis_pipeline/20201027_pileup_data/pileup_output/28s_pileup_table.csv\n",
      "Detected 28s as contig used in filename\n",
      "/Users/lij/PycharmProjects/eclip_analysis_pipeline/20201027_pileup_data/pileup_output/5s_pileup_table.csv\n",
      "Detected 5s as contig used in filename\n",
      "/Users/lij/PycharmProjects/eclip_analysis_pipeline/20201027_pileup_data/pileup_output/5.8s_pileup_table.csv\n",
      "Detected 5.8s as contig used in filename\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "import os\n",
    "\n",
    "#Human rRNA contigs and their lengths, as generated by Alex's fastawork.py\n",
    "contig_list = {'18s': 1869, '28s': 5070, '5.8s': 157, '5s': 121}\n",
    "\n",
    "def read_pileup_data(filepath, protein_list=None):\n",
    "    '''\n",
    "    Accepts filepath to .csv file containing pileup data (RPMS - reads per million) of a given contig, \n",
    "    and optionally, a list of protein names to search for in the dataset. \n",
    "    Outputs .csv file of pileup RPMs binned by nucleotide position in the contig\n",
    "    '''\n",
    "    global contig_list\n",
    "    \n",
    "    #Looks for contig listed in the filename\n",
    "    filename = filepath.split('/')[-1]\n",
    "    contig = filename.split('_')[0]\n",
    "    print(f\"Detected {contig} as contig used in filename\")\n",
    "    \n",
    "    #Read .csv file of pileup data\n",
    "    pileup_data = pd.read_csv(filepath,index_col=0)\n",
    "\n",
    "    #Cleans up column names by removing \"_sorted_CONTIG_[contig]\" trailer\n",
    "    pileup_data.columns = [re.search(\"(ENCSR[A-Za-z0-9]*_[A-Za-z0-9]*_rep[0-9]*)_sorted\",column).group(1) for column in pileup_data.columns]\n",
    "    \n",
    "    if protein_list != None:\n",
    "        #Filter dataset columns to only include those containing the proteins listed in \"protein_list\"\n",
    "        protein_regex = '_|'.join(protein_list)\n",
    "        filtered_pileup = pileup_data.filter(regex=protein_regex,axis=1)\n",
    "    else:\n",
    "        filtered_pileup=pileup_data\n",
    "\n",
    "    #Infers contig length by number of rows in the input pileup data. \n",
    "    #Checks it against the length listed in the global \"contig_list\" variable\n",
    "    contig_length = len(pileup_data.index)\n",
    "    assert contig_length == contig_list[contig], \"File row length does not match contig length\"\n",
    "    \n",
    "    #Generates 20nt bins and averages them to generate an output Dataframe of binned RPMs\n",
    "    groups = np.arange(1,contig_length+1,1/20)[0:contig_length]\n",
    "    groups = np.floor(groups)\n",
    "    pileup_group = filtered_pileup.groupby(groups)\n",
    "    avg_pileup = pileup_group.mean().round(0)\n",
    "    avg_pileup.index.name = \"20nt_group\"\n",
    "    avg_pileup.columns.name = f\"sample ({contig})\"\n",
    "\n",
    "    return avg_pileup\n",
    "    # #group by 20nt and average. Last group only contains 10nt\n",
    "    # groups = np.arange(1,95,1/20)[0:1870]\n",
    "    # groups = np.floor(groups)\n",
    "    # pileup_group = pileup_data.groupby(groups)\n",
    "    # avg_pileup = pileup_group.mean().round(0)\n",
    "    # avg_pileup.index.name = \"20nt_group\"\n",
    "    # avg_pileup.columns.name = \"sample\"\n",
    "    # avg_pileup.to_csv(\"./18S_average_pileup.csv\")\n",
    "\n",
    "    # with open(\"./18S_average_pileup.pkl\", \"wb\") as file:\n",
    "    #     pickle.dump(avg_pileup, file)\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    path = \"/Users/lij/PycharmProjects/eclip_analysis_pipeline/20201027_pileup_data/pileup_output/\"\n",
    "    file_list = os.listdir(path)\n",
    "    file_list = list(filter(lambda filename: filename.endswith(\"_pileup_table.csv\"), file_list))\n",
    "    path_file_list = [path+filename for filename in file_list]\n",
    "    \n",
    "    average_pileup_output_path = path +\"average_pileup_output/\"\n",
    "    if not os.path.exists(average_pileup_output_path):\n",
    "        os.mkdir(average_pileup_output_path)\n",
    "    \n",
    "    for filepath in path_file_list:\n",
    "        avg_pileup = read_pileup_data(filepath)\n",
    "        output_filepath = filepath.split('/')[-1].replace('_pileup_table.csv','_average_pileup.csv')\n",
    "        avg_pileup.to_csv(average_pileup_output_path + output_filepath)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "port sliding window code\n",
    "clear pileup_data column names\n",
    "drop first \"unamed 0\" column\n",
    "Figure out which proteins are controls using Anthony's list of accession numbers\n",
    "\n",
    "LONG-TERM:\n",
    "-Learn how to port Anthony's R code to Python\n",
    "\n",
    "Overview: eCLIP (for yourself - don't present)\n",
    "Pipeline - Anthony's slide\n",
    "Talk about filtering + pileup step (what did we filter, show equation on how reads per million were calculated)\n",
    "Stats (if we get there)\n",
    "Future directions"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:rosalind]",
   "language": "python",
   "name": "conda-env-rosalind-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
